include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -789,8 +789,10 @@ class EnableIterativeBase : public IterativeBase {
         std::shared_ptr<const stop::CriterionFactory> new_stop_factory) override
     {
         auto exec = self()->get_executor();
-        if (new_stop_factory && new_stop_factory->get_executor() != exec) {
-            new_stop_factory = gko::clone(exec, new_stop_factory);
+        if (new_stop_factory && new_stop_factory->get_executor() != exec &&
+            !exec->memory_accessible(new_stop_factory->get_executor())) {
+            GKO_INVALID_STATE(
+                "Cross-executor assignment of factories is currently broken.");
```

Comment:
The clone is not propagated to all internal factories, which leads to the inconsistencies. Consider the combined criterion, the factory for that stores the parameters, which are a std vector of criteria factories. The clone of the factory is implemented by its copy assignment. That assignment will just copy assign the stored parameters, so the executor of the internal factories is not changed. 
This already affects the iteration criterion, since it will try setting the convergence status using the criterion executor, which may not be the same executor as the convergence status.
TBH, I have no idea why that previously worked at all. Perhaps because now I'm checking more than one criteria?

Reply:
I would also like to investigate this further

Reply:
That's a good point, it most likely has to do with Combined not handling executors correctly, but we can fix that easily :)

Reply:
I manually created copy assignment for `Combined::Factory` that fixes it. But I'm not happy with this, since cloning factory seems to be inherently broken.

Reply:
Within the context of cloning a solver, it is fine. But it would break again if someone tries to clone a solver factory, or any other factory.

Reply:
If we copy to the correct executor only at stopping criterion generation, it should be fine, right?

Reply:
we don't require the parameters of a factory to share a common executors